core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java

```diff
@@ -0,0 +1,218 @@
+        // be created if the first write starts at offset 0. This allocates enough space for the
+        // write.
+        mWorker.createBlockRemote(sessionId, blockId, mStorageTierAssoc.getAlias(0), length);
+      } else {
+        // Allocate enough space in the existing temporary block for the write.
+        mWorker.requestSpace(sessionId, blockId, length);
+      }
+      writer = mWorker.getTempBlockWriterRemote(sessionId, blockId);
+      writer.append(buffer);
+
+      RPCBlockWriteResponse resp =
+          new RPCBlockWriteResponse(sessionId, blockId, offset, length, RPCResponse.Status.SUCCESS);
+      ChannelFuture future = ctx.writeAndFlush(resp);
+      future.addListener(ChannelFutureListener.CLOSE);
+      future.addListener(new ClosableResourceChannelListener(writer));
+    } catch (Exception e) {
+      LOG.error("Error writing remote block : {}", e.getMessage(), e);
+      RPCBlockWriteResponse resp =
+          RPCBlockWriteResponse.createErrorResponse(req, RPCResponse.Status.WRITE_ERROR);
+      ChannelFuture future = ctx.writeAndFlush(resp);
+      future.addListener(ChannelFutureListener.CLOSE);
+      if (writer != null) {
+        writer.close();
+      }
+    }
+  }
+
+  /**
+   * @return how much of a file to read. When {@code len} is {@code -1}, then
+   * {@code fileLength - offset} is used.
+   */
+  private long returnLength(final long offset, final long len, final long fileLength) {
+    return (len == -1) ? fileLength - offset : len;
+  }
+
+  /**
+   * Validates the bounds of the request. An uncaught exception will be thrown if an
+   * inconsistency occurs.
+   *
+   * @param req The initiating {@link RPCBlockReadRequest}
+   * @param fileLength The length of the block being read
+   */
+  private void validateBounds(final RPCBlockReadRequest req, final long fileLength) {
+    Preconditions.checkArgument(req.getOffset() <= fileLength,
+        "Offset(%s) is larger than file length(%s)", req.getOffset(), fileLength);
+    Preconditions.checkArgument(
+        req.getLength() == -1 || req.getOffset() + req.getLength() <= fileLength,
+        "Offset(%s) plus length(%s) is larger than file length(%s)", req.getOffset(),
+        req.getLength(), fileLength);
+  }
+
+  /**
+   * Returns the appropriate {@link DataBuffer} representing the data to send, depending on the
+   * configurable transfer type.
+   *
+   * @param req The initiating {@link RPCBlockReadRequest}
+   * @param reader The {@link BlockReader} for the block to read
+   * @param readLength The length, in bytes, of the data to read from the block
+   * @return a {@link DataBuffer} representing the data
+   * @throws IOException if an I/O error occurs when reading the data
+   * @throws IllegalArgumentException if TRANSFER is used and the data source is not a file channel
+   */
+  private DataBuffer getDataBuffer(RPCBlockReadRequest req, BlockReader reader, long readLength)
+      throws IOException, IllegalArgumentException {
```

Comment:
I think its better to document explicit unchecked exceptions, does it go against our current convention?